
<div style="text-align: center;">

  <!-- Logotipo -->
  <img src="img/LogoUesc.png" alt="UESC" style="width: 60px; display: block; margin: 0 auto 5px auto;">

  <!-- Cabeçalho -->
  <div style="font-size: 20px;"><strong>UNIVERSIDADE ESTADUAL DE SANTA CRUZ - UESC</strong></div>
  <div style="font-size: 18px;">DEPARTAMENTO DE ENGENHARIAS E COMPUTAÇÃO</div>
  <div style="font-size: 18px;">ENGENHARIA QUÍMICA</div>

  <hr style="margin: 20px 0; border: 1px solid #0074B7;">

  <!-- Título do Trabalho -->
  <div style="font-size: 22px; font-weight: bold; margin-bottom: 10px;">Introdução à Controle de Processos</div>
  <div style="font-size: 20px; font-weight: bold; margin-bottom: 10px;">Aplicação Computacional no Controle de um Separador Trifásico</div>

  <!-- Autor e orientador -->
  <div style="font-size: 18px;">Prof. Dr. E.R.Edwards</div>

<hr style="margin: 20px 0; border: 1px solid #0074B7;">
  <!-- Data -->
<!-- <div style="font-size: 16px; margin-top: 15px;">Ilhéus - BA, Abril de 2025</div> -->
</div>

__Ação Proporcinonal Discreta:__

A implementação digital da ação proporcional é direta. Partimos da expressão contínua:

$ 
u_p(t) = K_p.e(t)
$

que na forma discrreta, torna-se:

$
u_{p,k} = K_p.e_k
$

ou seja, basta multiplicar o erro atual pelo ganho proporciola.

__Aproximação Integral__

A integral continua é dada por:

$$
I_{(t)}= \int_0^t e(\tau)d(\tau)
$$

como o computador não calcula integrais analítiacas, utiliza-se uma aproximação numérica simples baseada na soma acumulada das áreas retangulares formuladas pelo erro em cada intervalo de tempo. Em cada passo, a área aproximada é $e_k \Delta t$, e a integral acumulada é atualizada por:

$$
I_{k+1} = I_k + e_k.\Delta t
$$

Essa equação é fundamental na implementação digital do controlador PI. Ela mostra que a integral nada mais é do que uma soma acumulada dos erros ao longo  do tempo.

__Interpretação Física__

No Separador Trifásico, se a interface permanece acima do _setpoint_, o erro será positivo. A cada intervalo de tempo o controlador acumula uma pequena parcela do erro, fazendo com que $I_k$ cresça continuamente.

À medida que a integral acumula, a ação do controlador também aumenta, abrindo mais a válvula de saída de água. Isso eleva a vazão $Q_{w, out}$ e reduz a altura da interface. É justamente esse acúmulo progressivo que permite o controlador PI eliminar o erro permanente. 

__Controlador PI Discreto__

Combinando os termos proporcional e integral, o controlador PI discreto é expresso por:

$
u_k = K_p.e_k + K_i.I_k
$

com a integral atualizada por:

$ 
I_{k+1} = I_k + e_k. \Delta t 
$ 

O algoritmo completo consiste em medir a interface, calcular o erro, atualizar a integral, determinar a ação de controle e ajustar a abertura da válvula.  Esse ciclo se repete continuamente ao longo da operação.























__Exemplo Numérico__

Considere os parâmetros:

$$  K_p = 2, \ \ \ K_i =  0,5 \ \ \ \Delta t = 1 \ h $$

e suponha que o erro inicial seja:

$$ e_o = 0.10 \ m $$

Sendo $k$ o contador.

Se a integral inicial é nula, $I_o = 0$, então:

$$ I_1 = I_o + e_o \Delta t  $$

$$ I_1 = 0 + (0.10) \times(1) = 0.10 $$

A ação de controle no instante seguinte será:

$$
2 \times (0.10) + 0.5 \times (0.10) = 0.25
$$ 

A ação total é composta por uma parcela proporcional e outra integral. No instante seguinte, enquanto existir erro, a integral continuará crescendo, reforçanco a ação de controle.

__Observação importante:__

A equação discreta da integral:

$ I_{k+1} I_k + e_k. \Delta t $

é a mesma utilizada nos programas em Python desenvolvidos anteriormente para o controle da interface água-óleo. Isso mostra que os algoritimos computacionais não são aproximações arbitrárias, mas sim a forma digital de um controlador PI industrial.

__Exercício Proposto:__

__Evolução da Ação Integral__

Considere o mesmo controlador PI apresentado anteriormente. Os parâmetros do controlador são: $K_p = 2$,  $K_i = 0,5$ e o passo de integral é: $\Delta t = 1 \ h$. Suponha que o erro permanece constante durante os primeiros instantes da operação: $e_k = 0.10 \ m$. A integral inicial é $I_o = 0$. A ação de controle é calculada por: 

$$ u_k = K_p.e_k + K_i.I_k  $$

Determine para os cinco primeiros passos de tempo:

1. O valor da integral ($I_k$)

2. O valor da ação de controle ($u_k$)

3. A contribuição proporcional;

4. A contribuição integral




Complete a tabela abaixo:

| k | $e_k$ | $I_k$ | $k_p e_k$ | $u_k$ |
|---|-------|-------|-----------|-------|
| 0 | 0.10 | 0.00 | 0.20 | ? |
| 1 | 0.10 | ?    | 0.20 | ? |
| 2 | 0.10 | ?    | 0.20 | ? |
| 3 | 0.10 | ?    | 0.20 | ? |
| 4 | 0.10 | ?    | 0.20 | ? |
| 5 | 0.10 | ?    | 0.20 | ? |

Pergunta:

1. O termo proporcional varia ao longo do tempo?
2. O termo integral varia ao longo do tempo ?
3. Qual parcela é responsável pelo aumento gradual da ação de controle ?
4. O que aconteceria se o erro permanecesse diferente de zero durante muito tempo?

In [9]:
# Solução 
# I_K+1 = I_k + e_k Delta_t]

# Dados de entrada (valores fornecidos no enunciado)
ek = 0.10
Delta_t = 1  # horas
Io = 0.00 

Kp = 2       # Ganho proporcional
Ki = 0.5     # Ganho integral

In [10]:
# Cálculo das integrais
I1 = Io + ek * Delta_t
I2 = I1 + ek * Delta_t
I3 = I2 + ek * Delta_t
I4 = I3 + ek * Delta_t
I5 = I4 + ek * Delta_t

# Imprimindo os valores das integrais:

def line():
    print("-"*30)

line()
print('Para K = 0 a Integral Io = ', Io)
print('Para K = 1 a Integral I1 = ', I1)
print('Para K = 2 a Integral I2 = ', I2)
print('Para K = 3 a Integral I3 = ', I3)
print('Para K = 4 a Integral I4 = ', I4)
print('Para K = 5 a Integral I5 = ', I5)

------------------------------
Para K = 0 a Integral Io =  0.0
Para K = 1 a Integral I1 =  0.1
Para K = 2 a Integral I2 =  0.2
Para K = 3 a Integral I3 =  0.30000000000000004
Para K = 4 a Integral I4 =  0.4
Para K = 5 a Integral I5 =  0.5


Cálculo da ação de controle.

In [ ]:
# uk = Kp.ek + Ki.Ik
uo = Kp * ek + Ki * Io
u1 = Kp * ek + Ki * I1
u2 = Kp * ek + Ki * I2
u3 = Kp * ek + Ki * I3
u4 = Kp * ek + Ki * I4
u5 = Kp * ek + Ki * I5


# Imprimindo os valores da ação de controle, temos:

def line():
    print("-"*30)

line()
print('Para u0 =', uo)
print('Para u1 =', u1)
print('Para u2 =', u2)
print('Para u3 =', u3)
print('Para u4 =', u4)
print('Para u5 =', u5)


------------------------------
Para u0 = 0.2
Para u1 = 0.25
Para u2 = 0.30000000000000004
Para u3 = 0.35000000000000003
Para u4 = 0.4
Para u5 = 0.45


: 

Completando a Tabela:

| k | $e_k$ | $I_k$ | $k_p e_k$ | $u_k$ |
|---|-------|-------|-----------|-------|
| 0 | 0.10 | 0.00 | 0.20 | 0.20 |
| 1 | 0.10 | 0.10 | 0.20 | 0.25 |
| 2 | 0.10 | 0.20 | 0.20 | 0.30 |
| 3 | 0.10 | 0.30 | 0.20 | 0.35 |
| 4 | 0.10 | 0.40 | 0.20 | 0.40 |
| 5 | 0.10 | 0.50 | 0.20 | 0.45 |

Observa-se que a parcela proporcional permanece constante, enquanto a parcela integral cresce continuamente.

=> Essa é exatamente a característica que permite ao controlador PI eliminar o erro  permanente observado em Processos Industriais.